posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+Started @ offset: 329
+Started @ offset: 423
+Started @ offset: 376
+Started @ offset: 470
+Started @ offset: 517
+[success] Total time: 1 s, completed 19 Apr 2021, 18:51:35
+Started @ offset: 564
+Started @ offset: 611
+Started @ offset: 658
+Started @ offset: 705
+Started @ offset: 752
+Started @ offset: 799
+Started @ offset: 846
+Started @ offset: 893
+Started @ offset: 940
+Started @ offset: 987
+Started @ offset: 1034
+Started @ offset: 1081
+Started @ offset: 1128
+Started @ offset: 1175
+Started @ offset: 1222
+Started @ offset: 1269
+Started @ offset: 1316
+Started @ offset: 1363
+Started @ offset: 1410
+Started @ offset: 1457
+Started @ offset: 1504
+Started @ offset: 1551
+Started @ offset: 1598
+Started @ offset: 1645
+Started @ offset: 1692
+Started @ offset: 1739
+Started @ offset: 1786
+Started @ offset: 1833
+Started @ offset: 1880
+Started @ offset: 1927
+Started @ offset: 1974
+Started @ offset: 2021
+Started @ offset: 2068
+Started @ offset: 2115
+Started @ offset: 2162
+Started @ offset: 2209
+Started @ offset: 2256
+Started @ offset: 2303
+Started @ offset: 2350
+Started @ offset: 2397
+Started @ offset: 2444
+Started @ offset: 2491
+Started @ offset: 2538
+Started @ offset: 2585
+Started @ offset: 2632
+Started @ offset: 2679
+Started @ offset: 2726
+Started @ offset: 2773
+Started @ offset: 2820
+Started @ offset: 2867
+Started @ offset: 2914
+Started @ offset: 2961
+Pi:      3.141592653589793238462643383279502884197169...
+|Delta|: 2.8076968E-3630
+Calculation time: 1.345
+
+----
+
+What we can observe is that, with 3,008 terms, we have correctly calculated more
+than 3,600 digits accurately.
+
+If we set the size of the ForkJoinPool to 1, 2, 4, 8, 12, 16, 32 we get the
+following values for the calculation time (in seconds) respectively: 12.12,
+6.23, 3.03, 1.78, 1.34, 1.36, 1.43
+
+So, we see a near linear speed-up by going from one thread to two threads.
+A further increase of the thread-count doesn't yield a further linear speed-up:
+this may be caused by different factors, not in the least by the fact that we
+have a single chip processor with a shared on-chip cache.
+Of course, since it's a six core CPU (with hyper-threads that don't yield the
+same performance as the regular CPU threads), we don't get a speed-up beyond 12
+threads in the ForkJoinPool.
+
+== Conclusion on the `Future` approach
+
+Scala's `Future` API presents a very powerful way to perform asynchronous and
+concurrent execution of code. Even though it may take some time to wrap one's
+head around it, once you grasp it, it's pretty cool and very powerful.
+
+Now, as for π, is the approach used in this article to calculate π a realistic way
+to calculate this number to say multi-billion digit precision? Not really for
+multiple reasons.
+
+First of all, this algorithm runs in-memory. If we consider that the current
+record for calculating π digits is at 50 trillion digits, there's no computer
+that can hold the required size of numbers in memory.
```

Comment:
Why won't this fit into memory? Not in string representation perhaps, but surely it should fit into a very big integer?

Reply:
Ok so this might be many gigabytes per number, not really feasible... not even wolfram alpha wants to tell me how many bits you need for that number.

Reply:
Indeed